### PR TITLE
[Platform Compatibility] Added CGROUP HEIRACHY and Systemd  Checks

### PR DESCRIPTION
**Pull request by nimanch**
Added additional checks for
1. cgroup hierarchy : Checks whether control group hierarchy is present. This is to ensure a container engine can run on the target device 
2. systemd check : Checks whether or not systemd is present. If systemd is not present.
3. Additionally added some other helper functions.

Todo's
2. Add more required kernel config checks for running a OCI Compliant container enginer after further analysis
3. Finalize wording for the check based on UX Flow

Testing
1. Test on OpenSUSE Linux
image
2. Test on Ubuntu 20.04 without systemd
image
3. Test on OpenWRT Image
image

## Azure IoT Edge PR checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines and Best Practices
- [x] I have read the [contribution guidelines](https://github.com/azure/iotedge#contributing).
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request includes a concise summary of the enhancement or bug fix.

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.
- Description of the pull request includes 
	- [x] concise summary of tests added/modified
	- [x] local testing done.  

### Draft PRs
- Open the PR in `Draft` mode if it is:
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.

_Note: We use the kodiakhq bot to merge PRs once the necessary checks and approvals are in place. When it merges a PR, kodiakhq converts the PR title to the commit title, PR description to the commit description, and squashes all the commits in the PR to a single commit. The net effect is that entire PR becomes a single commit. Please follow the best practices mentioned [here](https://chris.beams.io/posts/git-commit/#:~:text=The%20seven%20rules%20of%20a%20great%20Git%20commit,what%20and%20why%20vs.%20how%20For%20example%3A%20) for the PR title and description_

### CODE REVIEW

File: platform-validation/scripts/compatibility.sh [hunk 1]
```diff
@@ -387,42 +422,80 @@ perform_capability_check_container(){
...
     fi
-    wrap_pass "capability_check_container" "Pass"
+    wrap_pass "check_net_cap_bind_container" "Pass"
+}
+
+# bits of this were adapted from moby check-config.shells
+# See https://github.com/moby/moby/blob/master/contrib/check-config.sh
+
+check_cgroup_heirachy()
+{
```
Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:435]: Can you link the cgroup GH issue  here for reference please?

Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:435]: nit: curly braces to the end of the func definition just like other functions

File: platform-validation/scripts/compatibility.sh [hunk 2]
```diff
@@ -387,42 +422,80 @@ perform_capability_check_container(){
...
+    if [ $EXITCODE -eq 0 ]; then
+        wrap_pass "check_cgroup_heirachy"
+    else
+        wrap_fail "check_cgroup_heirachy"
+    fi
+}
+
+
+check_systemd(){
+
```
Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:480]: wrap_debug for letting the Cx know that this check is getting performed...

File: platform-validation/scripts/compatibility.sh [hunk 3]
```diff
@@ -323,16 +359,15 @@ perform_cleanup(){
...
 # ------------------------------------------------------------------------------
-# Check whether the Target Device can be used to set capability. EdgeHub Runtime component sets CAP_NET_BIND which is Required
-# for Azure IoT Edge Operation
+# Check whether the Target Device can be used to set capability. EdgeHub   Runtime component sets CAP_NET_BIND which is Required for Azure IoT Edge Operation.
 # ------------------------------------------------------------------------------
-perform_capability_check_host(){
+check_net_cap_bind_host(){
     wrap_debug "Setting the CAP_NET_BIND_SERVICE capability on the host..."
     
       # Check dependencies
```
Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:367]: nit: fix indentation

File: platform-validation/scripts/compatibility.sh [hunk 4]
```diff
@@ -12,6 +12,30 @@ ARCH=""
 #  Text Formatting
 #  Derived from : https://github.com/moby/moby/blob/master/contrib/check-config.sh
 # ------------------------------------------------------------------------------
+
+possibleConfigs="
```
Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:16]: variables names are all have to be uppercase. can you change it to `POSSIBLE_CONFIGS`?

File: platform-validation/scripts/compatibility.sh [hunk 5]
```diff
@@ -12,6 +12,30 @@ ARCH=""
...
+	: "${CONFIG:=/proc/config.gz}"
+fi
+
+if ! command -v zgrep > /dev/null 2>&1; then
+	zgrep() {
+		zcat "$2" | grep "$1"
+	}
+fi
+
+is_set() {
```
Comment (ggjjj) [inline platform-validation/scripts/compatibility.sh:35]: Do you plan to use this?

Comment (nimanch) [reply platform-validation/scripts/compatibility.sh:35]: I think I need some follow up investigation, before I can start using them. But yes, i do intend to use them

**Pull request merged**